# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webgpu/ipc/WebGPUChild.cpp
# Commit: dfd34fb1aade
# Full Hash: dfd34fb1aade1295579715b1c21f1e51bbfe370e
# Author: Dzmitry Malyshau <dmalyshau@mozilla.com>
# Date: 2021-05-18 21:36:28
# Regressor Bug: 1709462
# File Overlap Count: 2
# Description:
#   Bug 1709462 - Fix freeing of WebGPU bind group layouts r=jimb
#   
#   When the pipeline layout is implicit at pipeline creation,
#   we collect the IDs of BGLs, to be able to produce GPUBindGroupLayout
#   object upon user request. However, the produced object didn't know if
# ==============================================================================

diff -r 5c6f0950714d -r dfd34fb1aade dom/webgpu/ipc/WebGPUChild.cpp
--- a/dom/webgpu/ipc/WebGPUChild.cpp	Tue May 18 16:24:49 2021 +0000
+++ b/dom/webgpu/ipc/WebGPUChild.cpp	Tue May 18 16:26:27 2021 +0000
@@ -583,6 +583,7 @@
 
 RawId WebGPUChild::DeviceCreateComputePipeline(
     RawId aSelfId, const dom::GPUComputePipelineDescriptor& aDesc,
+    RawId* const aImplicitPipelineLayoutId,
     nsTArray<RawId>* const aImplicitBindGroupLayoutIds) {
   ffi::WGPUComputePipelineDescriptor desc = {};
   nsCString label, entryPoint;
@@ -600,7 +601,8 @@
   ByteBuf bb;
   RawId implicit_bgl_ids[WGPUMAX_BIND_GROUPS] = {};
   RawId id = ffi::wgpu_client_create_compute_pipeline(
-      mClient, aSelfId, &desc, ToFFI(&bb), implicit_bgl_ids);
+      mClient, aSelfId, &desc, ToFFI(&bb), aImplicitPipelineLayoutId,
+      implicit_bgl_ids);
 
   for (const auto& cur : implicit_bgl_ids) {
     if (!cur) break;
@@ -658,6 +660,7 @@
 
 RawId WebGPUChild::DeviceCreateRenderPipeline(
     RawId aSelfId, const dom::GPURenderPipelineDescriptor& aDesc,
+    RawId* const aImplicitPipelineLayoutId,
     nsTArray<RawId>* const aImplicitBindGroupLayoutIds) {
   // A bunch of stack locals that we can have pointers into
   nsTArray<ffi::WGPUVertexBufferLayout> vertexBuffers;
@@ -772,7 +775,8 @@
   ByteBuf bb;
   RawId implicit_bgl_ids[WGPUMAX_BIND_GROUPS] = {};
   RawId id = ffi::wgpu_client_create_render_pipeline(
-      mClient, aSelfId, &desc, ToFFI(&bb), implicit_bgl_ids);
+      mClient, aSelfId, &desc, ToFFI(&bb), aImplicitPipelineLayoutId,
+      implicit_bgl_ids);
 
   for (const auto& cur : implicit_bgl_ids) {
     if (!cur) break;